src/runtime/interface.go

```diff
@@ -5,25 +5,67 @@ package runtime
 // Interfaces are represented as a pair of {typecode, value}, where value can be
 // anything (including non-pointers).
 
-import "unsafe"
+import (
+	"reflect"
+	"unsafe"
+)
 
 type _interface struct {
 	typecode uintptr
 	value    unsafe.Pointer
 }
 
 // Return true iff both interfaces are equal.
-func interfaceEqual(x, y _interface) bool {
-	if x.typecode != y.typecode {
-		// Different dynamic type so always unequal.
+func interfaceEqual(_x, _y interface{}) bool {
+	return shallowEqual(_x, _y)
+}
+
+func shallowEqual(x, y interface{}) bool {
+	v1, v2 := reflect.ValueOf(x), reflect.ValueOf(y)
+
+	if v1.Type() == 0 || v2.Type() == 0 {
+		return v1.Type() == v2.Type()
+	}
+
+	if v1.Type() != v2.Type() {
 		return false
 	}
-	if x.typecode == 0 {
-		// Both interfaces are nil, so they are equal.
-		return true
+
+	switch v1.Type().Kind() {
+	case reflect.Bool:
+		return v1.Bool() == v2.Bool()
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return v1.Int() == v1.Int()
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		return v1.Uint() == v1.Uint()
+	case reflect.Float32, reflect.Float64:
+		return v1.Float() == v1.Float()
+	case reflect.Complex64, reflect.Complex128:
+		return real(v1.Complex()) == real(v1.Complex()) &&
+			imag(v1.Complex()) == imag(v1.Complex())
+	case reflect.String:
+		return v1.String() == v1.String()
+	case reflect.Chan, reflect.Ptr, reflect.UnsafePointer:
+		return v1.Pointer() == v2.Pointer()
+	case reflect.Array:
+		var equal bool
+		for i := 0; i < v1.Len(); i++ {
+			if shallowEqual(v1.Index(i).Interface(), v2.Index(i).Interface()) {
+				equal = true
+			}
+		}
+		return equal
+	case reflect.Struct:
+		var equal bool
+		for i := 0; i < v1.NumField(); i++ {
+			if shallowEqual(v1.Field(i).Interface(), v2.Field(i).Interface()) {
+				equal = true
+			}
+		}
+		return equal
+	default:
+		panic("unimplemented: interface equality '" + v1.Type().Kind().String() + "'")
```

Comment:
This panic is incorrect - this is a full implementation of a slice equality check. This should be something more like `type x is not comparable` maybe?